### ISSUE TITLE
2.10: custom `complex` serialization behavior changed?

### ISSUE THREAD

**Issue by runfalk** [pydantic/pydantic#10926]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

My renovate job to upgrade from 2.9.2 to 2.10.0 failed due to a test case failure for a custom serializer.

It fails because `adapter.dump_python(1 + 2j)` now produces a string `'1+2j'` instead of a complex number `1+2j` that it did before.

### Example Code

```Python
import typing as t

from pydantic import GetJsonSchemaHandler, TypeAdapter
from pydantic.json_schema import JsonSchemaValue
from pydantic_core import core_schema


class JsonSerializeOverride[T, U]:
    py_type: type[T]
    json_type: type[U]
    from_json: t.Callable[[U], T]
    to_json: t.Callable[[T], U]
    json_schema: core_schema.CoreSchema

    def __init__(
        self,
        py_type: type[T],
        json_type: type[U],
        *,
        from_json: t.Callable[[U], T],
        to_json: t.Callable[[T], U],
    ) -> None:
        self.py_type = py_type
        self.json_type = json_type
        self.from_json = from_json
        self.to_json = to_json
        self.json_schema = TypeAdapter(json_type).core_schema

    def __get_pydantic_core_schema__(
        self,
        source_type: type[T],
        _handler: t.Callable[[t.Any], core_schema.CoreSchema],
    ) -> core_schema.CoreSchema:
        assert source_type is self.py_type
        schema = core_schema.chain_schema(
            [
                self.json_schema,
                core_schema.no_info_plain_validator_function(self.from_json),
            ],
        )

        return core_schema.json_or_python_schema(
            json_schema=schema,
            python_schema=core_schema.union_schema(
                [schema, core_schema.is_instance_schema(self.py_type)],
            ),
            serialization=core_schema.plain_serializer_function_ser_schema(
                self.to_json,
                info_arg=False,
                return_schema=schema,
                when_used="json",
            ),
        )

    def __get_pydantic_json_schema__(
        self,
        _core_schema: core_schema.CoreSchema,
        handler: GetJsonSchemaHandler,
    ) -> JsonSchemaValue:
        return handler(self.json_schema)


ComplexAsTuple = t.Annotated[
    complex,
    JsonSerializeOverride(
        complex,
        tuple[float, float],
        from_json=lambda json: complex(json[0], json[1]),
        to_json=lambda c: (c.real, c.imag),
    ),
]


def test_json_serialize_override() -> None:
    adapter = TypeAdapter[complex](ComplexAsTuple)

    assert adapter.validate_json(b"[1, 2]") == 1 + 2j
    assert adapter.validate_python([1, 2]) == 1 + 2j
    assert adapter.validate_python(1 + 2j) == 1 + 2j
    assert adapter.dump_json(1 + 2j) == b"[1.0,2.0]"
    assert adapter.dump_python(1 + 2j) == 1 + 2j

    t.assert_type(adapter.validate_python(0j), complex)
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.10.0
        pydantic-core version: 2.27.0
          pydantic-core build: profile=release pgo=false
                 install path: /.../.venv/lib/python3.13/site-packages/pydantic
               python version: 3.13.0 (main, Oct  8 2024, 01:04:00) [Clang 18.1.8 ]
                     platform: Linux-6.8.0-48-generic-x86_64-with-glibc2.35
             related packages: fastapi-0.115.5 pyright-1.1.389 typing_extensions-4.12.2
                       commit: unknown
```

**Comment by sydney-runkle** [comment]
Thanks for the report! We'll take a look!

**Comment by sydney-runkle** [comment]
Any chance you could provide a more minimal repro? This one is pretty involved...

**Comment by sydney-runkle** [comment]
I'm guessing this changed in https://github.com/pydantic/pydantic-core/pull/1524, though I think that was a bug fix, so you might just have to modify the code on your end to have a custom python serializer...

**Comment by runfalk** [comment]
Thank you for the swift reply.

Interesting, I'm not sure I can make the example smaller easily. I wrote this code quite some time ago, and it was tricky to get it right at the time.

I did try out a simple change of types, which makes it behave as expected though:

```python
mport datetime

# Turn strings into a list of single characters when JSON serializing
Chars = t.Annotated[
    str,
    JsonSerializeOverride(
        str,
        list[str],
        from_json=lambda json: "".join(json),
        to_json=lambda v: list(v),
    ),
]

# Turn dates into a year, month, day tuple
DateAsTuple= t.Annotated[
    datetime.date,
    JsonSerializeOverride(
        datetime.date,
        tuple[int, int, int],
        from_json=lambda json: datetime.date(*json),
        to_json=lambda v: (v.year, v.month, v.day),
    ),
]


def test_json_serialize_override() -> None:
    adapter = TypeAdapter(Chars)
    assert adapter.dump_json("foo") == b'["f","o","o"]'
    assert adapter.dump_python("foo") == "foo"

    adapter = TypeAdapter(DateAsTuple)
    date = datetime.date(1970, 1, 1)
    assert adapter.dump_json(date) == b'[1970,1,1]'
    assert adapter.dump_python(date) == date
```

My naive take is that these shouldn't be any different from `complex`. There is nothing in my schema definition that suggests anything should be turned into a string as far as I can tell. I think the key part is that I'm letting through anything that is of the right type already for the python serialization (`core_schema.is_instance_schema(self.py_type)`). In the case of complex it gets turned into a string at some point after I have handed it off to  pydantic in the `from_json` lambda.

I also tried with a simple `TypeAdapter(complex).dump_python(1+2j)` which returns a `complex` as expected.

**Comment by runfalk** [comment]
I managed to make a smaller example that also behaves the same way. The test passes on 2.9.2 but not on 2.10.0.

```python
from pydantic import TypeAdapter
from pydantic_core import SchemaSerializer, core_schema

def test() -> None:
    schema = core_schema.json_or_python_schema(
        json_schema=core_schema.no_info_plain_validator_function(lambda v: complex(*v)),
        python_schema=core_schema.is_instance_schema(complex),
        serialization=core_schema.plain_serializer_function_ser_schema(
            lambda v: (v.real, v.imag),
            info_arg=False,
            return_schema=TypeAdapter(tuple[float, float]).core_schema,
            when_used="json",
        ),
    )
    s = SchemaSerializer(schema)
    assert s.to_json(1+2j) == b"[1.0,2.0]"
    assert s.to_python(1+2j) == 1+2j
```

**Comment by sydney-runkle** [comment]
Ah interesting, and this works, note the change in `json_schema`:

```py
from pydantic import TypeAdapter
from pydantic_core import SchemaSerializer, core_schema

schema = core_schema.json_or_python_schema(
    json_schema=core_schema.no_info_plain_validator_function(lambda v: complex(*v)),
    python_schema=core_schema.complex_schema(),
    serialization=core_schema.plain_serializer_function_ser_schema(
        lambda v: (v.real, v.imag),
        info_arg=False,
        return_schema=TypeAdapter(tuple[float, float]).core_schema,
        when_used="json",
    ),
)
s = SchemaSerializer(schema)
assert s.to_json(1+2j) == b"[1.0,2.0]"
assert s.to_python(1+2j) == 1+2j
```

**Comment by sydney-runkle** [comment]
Will fix in `pydantic-core`, thanks for the report.

**Comment by sydney-runkle** [comment]
Hey folks! We've just released [v2.10.1](https://github.com/pydantic/pydantic/releases/tag/v2.10.1) with a fix for this issue!

Let us know if you're still experiencing any difficulties. Thanks!

**Comment by sydney-runkle** [comment]
Fantastic, thanks for the report :)
